docs/zephyr_02_functional_requirements.sdoc

```diff
@@ -1317,6 +1317,248 @@ As a Zephyr OS user I want to... ???,
+
+- unblock the waiting thread indicating the Take request was successful; and
+- indicate to the Giving thread that the Give request was successful.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-137
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Give without pending Take increases count by 1
+STATEMENT: >>>
+While no thread is blocked waiting to Take from this Counting Semaphore,
+
+When the Counting Semaphore Give is requested,
+
+the Counting Semaphore shall do all of the following:
+
+- increase the count of this semaphore by 1, up to the configured limit of this Semaphore; and
+- indicate to the Giving thread that the Give request was successful.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-138
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Give Request Invalid Parameter is Catastrophic Error
+STATEMENT: >>>
+When the Counting Semaphore Give is requested,
+
+If the Counting Semaphore Give parameters are invalid,
+
+Zephyr shall report a catastrophic failure.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-139
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Take when Count is >0
+STATEMENT: >>>
+While the Count of this Counting Semaphore is > 0,
+
+When the Counting Semaphore Take is requested,
+
+the Counting Semaphore shall do all of the following:
+
+- atomically reduce the Count of this Counting Semaphore by 1; and
+- indicate to the Taking thread that the Take request was successful.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-140
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Take blocked when Count is 0
+STATEMENT: >>>
+While the Count of this Counting Semaphore is 0,
+
+When the Counting Semaphore Take is requested,
+
+the Counting Semaphore shall block the thread requesting a Counting Semaphore Take.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-141
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphore
+TITLE: Counting Semaphore Take times out
+STATEMENT: >>>
+When a Counting Semaphore Take request timeout has expired,
+
+the Counting Semaphore shall unblock the waiting thread indicating the Take request failed due to a timeout.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-142
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Take Request Invalid Parameter is Catastrophic Error
+STATEMENT: >>>
+When the Counting Semaphore Take is requested,
+
+If the Counting Semaphore Take parameters are invalid,
+
+Zephyr shall report a catastrophic failure.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-143
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Take Request From Interrupt Context with Non-Immediate Timeout is Catastrophic Error
+STATEMENT: >>>
+When the Counting Semaphore Take is requested from Interrupt Context,
+
+If the Counting Semaphore Take timeout is non-immediate,
+
+Zephyr shall report a catastrophic failure.
+<<<
```

Comment:
Allowing a Semaphore Take in Interrupt context where Timeout is supported is inherently unsafe. It will cause a deadlock. Exchanging the deadlock for a catastrophic failure is not any safer.  And there is the issue of blocking and unblocking threads. Not all RTOSs will do that from inside an interrupt.  
Not sure how to react to this requirement except with great apprehension.

Reply:
This is a great time to figure out what users require. ;-)

BTW, the current implementation asserts if called from interrupt context. I'm not sure what it does when asserts are disabled.

Reply:
Greg is right though, current API docs explicitly allow to take a semaphore from ISR context under the constraint that the timeout has to be [K_NO_WAIT](https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/sem.c#L136)

Reply:
of course Greg is right, this is not the only place where it happens and where ASSERTs are using to perform error checking. This has been the design of Zephyr since the first day: enable asserts during development, get your code in order, then disable asserts and ship it. We then introduced checkif macros to turn those into runtime checks and still maintain footprint under control, we also introduced userspace with its own error checking and input validation stubs. 

Bottom line, in my previous comments here I am not disagreeing to error checking and input validation, but we need figure out how to specify it effectively in the requirements without going through each error checking in the code and creating a requirement for it, this way we might actually missing some very hot spots, because the requirements are just describing the existing code logic and not the implemented concept.

Reply:
https://docs.zephyrproject.org/latest/kernel/usermode/overview.html#high-level-policy-details

> For performance and footprint reasons Zephyr normally does little or no parameter error checking for kernel object or device driver APIs. Access from user mode through system calls involves an extra layer of handler functions, which are expected to rigorously validate access permissions and type of the object, check the validity of other parameters through bounds checking or other means, and verify proper read/write access to any memory buffers involved.

We have something, but here we are trying to deal with semaphores and keep the discussion focused on semaphores, we will get to the above and other requirements.

Reply:
Performance and footprint are at odds with defense-in-depth. For Zephyr the Semaphore API is a public interface. To minimize risks during usage and integration into a device all input data should be validated by default. This is a secure design principle echoed in ETSI 303 645 Provision 5.13-1.

AFAICT security trumps performance and footprint (for default behavior). I am fine with giving the Product Manufacturers the _option_ to disable parameter checking of a public API level. They are the only ones with the knowledge about the entire executable to decide when that is safe to do.

> here we are trying to deal with semaphores and keep the discussion focused on semaphores, 

Agreed. This appears because Semaphores can be acted upon from Interrupt, Supervisor, and User CPU contexts/modes and the Semaphore API is public. We must capture the required behavior. If this document is capturing requirements for the opaque Zephyr Kernel then the observable behavior may be to _exhibit_ a fatal error. If this document is capturing requirements for the opaque Zephyr Counting Semaphore within the transparent Zephyr Kernel then the observable behavior may be to _report_ a fatal error to the Zephyr Fatal Error Handler.

If you are already anticipating this will be covered at a higher layer requirement then I think we are getting ahead of ourselves. Eventually we will recognize that:

1. the purpose of tests is to verify the code-under-test fully meets the _requirements_ delegated to it,
2. the purpose of code coverage is to verify the code-under-test does nothing else,
3. there are more levels than High/Low.

I think it is wiser to capture everything for the Counting Semaphore sub-Element (e.g., kernel/sem.c, zephyr/kernel.h::k_sem*) before we move up a level to the opaque requirements on Zephyr Kernel. This pattern of moving up a decomposition level will re-occur with `arch`, `hal*`, executables, sources vs tools vs artifacts, etc.